crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,419 @@
+      prop = Proposal.where(barclamp: bc, name: inst).first
+      prop["deployment"][bc]["crowbar-queued"] = true
+      prop.save
+      logger.debug("queue proposal: exit #{inst} #{bc}")
+      [ delay, pre_cached_nodes ]
+    end
+
+    # Locking wrapper around dequeue_proposal_no_lock
+    def dequeue_proposal(bc, inst)
+      logger.debug("dequeue proposal: enter #{inst} #{bc}")
+      ret = false
+      begin
+        f = file_lock.acquire("queue", logger: logger)
+
+        if proposal_queue.empty?
+          logger.debug("dequeue proposal: exit #{inst} #{bc}: no entry")
+          return [200, {}]
+        end
+
+        dequeue_proposal_no_lock(bc, inst)
+      rescue StandardError => e
+        logger.error("Error dequeuing proposal for #{bc}:#{inst}: #{e.message} #{e.backtrace.join("\n")}")
+        logger.debug("dequeue proposal: exit #{inst} #{bc}: error")
+        return [400, e.message]
+      ensure
+        file_lock.release(f)
+      end
+      logger.debug("dequeue proposal: exit #{inst} #{bc}")
+      return dequeued ? [200, {}] : [400, '']
+    end
+
+    # Deps are satisfied if all exist, have been deployed and are not in the queue ATM.
+    def dependencies_satisfied?(deps)
+      item["deps"].all? do |dep|
+        depprop = Proposal.where(barclamp: dep["barclamp"], name: dep["inst"]).first
+        depprop_queued   = depprop["deployment"][dep["barclamp"]]["crowbar-queued"] rescue false
+        depprop_deployed = (depprop["deployment"][dep["barclamp"]]["crowbar-status"] == "success") rescue false
+
+        depprop && !depprop_queued && depprop_deployed
+      end
+    end
+
+    #
+    # NOTE: If dependencies don't form a DAG (Directed Acyclic Graph) then we have a problem
+    # with our dependency algorithm
+    #
+    def process_queue
+      logger.debug("process queue: enter")
+      loop_again = true
+      while loop_again
+        loop_again = false
+
+        # List contains proposals in the queue that can be applied
+        # Remove list contains proposals that were either deleted
+        # or should be re-queued?
+        # Proposals which reference non-ready nodes are also skipped.
+        list = []
+        begin
+          f = file_lock.acquire("queue", logger: logger)
+
+          if proposal_queue.empty?
+            logger.debug("process queue: exit: empty queue")
+            return
+          end
+          logger.debug("process queue: queue: #{queue.inspect}")
+
+          # Test for ready
+          remove_list = []
+          proposal_queue.proposals.each do |item|
+            prop = Proposal.where(barclamp: item["barclamp"], name: item["inst"]).first
+            if prop.nil?
+              remove_list << item
+              next
+            end
+
+            next unless dependencies_satisfied?(item)
+
+            nodes_map = elements_to_nodes_to_roles_map(prop["deployment"][item["barclamp"]]["elements"])
+            delay, pre_cached_nodes = elements_not_ready(nodes_map.keys)
```

Comment:
Lint/UselessAssignment: Useless assignment to variable - `pre_cached_nodes`. Use `_` or `_pre_cached_nodes` as a variable name to indicate that it won't be used.